starlette/testclient.py

```diff
@@ -514,63 +518,59 @@ def __enter__(self) -> "TestClient":
             def reset_portal() -> None:
                 self.portal = None
 
-            self.stream_send = StapledObjectStream(
-                *anyio.create_memory_object_stream(math.inf)
-            )
-            self.stream_receive = StapledObjectStream(
-                *anyio.create_memory_object_stream(math.inf)
+            shutdown_event = portal.call(anyio.Event)
+            fut, task_status = portal.start_task(self._lifespan, shutdown_event)
+
+            @stack.push
+            def wait_shutdown(*exc_info: object) -> bool:
+                portal.call(shutdown_event.set)
+                if exc_info != (None, None, None):
+                    fut.cancel()
+                else:
+                    fut.result()
+                return False
+
+            assert task_status["type"] in (
+                "lifespan.startup.complete",
+                "lifespan.startup.failed",
             )
-            self.task = portal.start_task_soon(self.lifespan)
-            portal.call(self.wait_startup)
-
-            @stack.callback
-            def wait_shutdown() -> None:
-                portal.call(self.wait_shutdown)
 
             self.exit_stack = stack.pop_all()
 
         return self
 
-    def __exit__(self, *args: typing.Any) -> None:
-        self.exit_stack.close()
+    def __exit__(self, *args: typing.Any) -> bool:  # type: ignore[override]
+        return self.exit_stack.__exit__(*args)
 
-    async def lifespan(self) -> None:
-        scope = {"type": "lifespan"}
-        try:
-            await self.app(scope, self.stream_receive.receive, self.stream_send.send)
-        finally:
-            await self.stream_send.send(None)
-
-    async def wait_startup(self) -> None:
-        await self.stream_receive.send({"type": "lifespan.startup"})
-
-        async def receive() -> typing.Any:
-            message = await self.stream_send.receive()
-            if message is None:
-                self.task.result()
-            return message
+    async def _lifespan(
+        self, shutdown_event: anyio.Event, *, task_status: anyio.abc.TaskStatus
+    ) -> None:
+        failures = []
```

Comment:
I don't like this - I think this should be a MemoryObjectStream and a task consuming errors and raising them

I can't raise the assertion directly in `async def send(` because it might not be run in the current task

Reply:
but I like having no concurrency in lifespan itself